[STORM-1336] - Evalute/Port JStorm cgroup support and implement cgroup support for RAS

jerrypeng opened the pull request:
Documentation in PR

https://github.com/apache/storm/pull/1054

revans2 commented:
Rat is complaining about some of your files not having the correct license.

revans2 commented on conf/defaults.yaml:
```diff
@@ -281,3 +281,16 @@ pacemaker.thread.timeout: 10
+# Also determines whether the unit tests for cgroup runs.  If cgroup.enable is set to false the unit tests for cgroups will not run
+cgroup.enable: false
+cgroup.supervisor.rootdir: "storm"
+cgroup.cgexec.cmd: "/bin/cgexec"
```
Can we please prefix these configs with "storm."  and probably also "storm.supervisor."

jerrypeng replied:
will fix

revans2 commented on storm-core/src/clj/org/apache/storm/util.clj:
```diff
@@ -541,6 +541,7 @@
     (.redirectErrorStream builder true)
     (doseq [[k v] environment]
       (.put process-env k v))
+    (log-message "Starting process with command: " command)
```
Can we make this debug or remove it if debugging is done?

jerrypeng replied:
will remove

revans2 commented on storm-core/src/jvm/org/apache/storm/Config.java:
```diff
@@ -2359,4 +2417,24 @@ public void setTopologyStrategy(Class<? extends IStrategy> clazz) {
         }
     }
+
+    public static String getCgroupRootDir(Map conf) {
```
Should we capitalize the G like in the other methods?

jerrypeng replied:
will change

jerrypeng replied:
fixed

revans2 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCenter.java:
```diff
@@ -0,0 +1,238 @@
+import java.util.Map;
+import java.util.Set;
+
+public class CgroupCenter implements CgroupOperation {
```
Again do we want all of these to be CGroup or Cgroup.  I don't really care so long as we are consistent everywhere.

jerrypeng replied:
will change to Cgroup

jerrypeng replied:
fixed

revans2 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCenter.java:
```diff
@@ -0,0 +1,238 @@
+        } catch (Exception e) {
+            LOG.error("Get hierarchies error", e);
+        } finally {
+            CgroupUtils.close(reader, br);
```
Can we update this to be java 7 like using the Closable and the try.

revans2 replied:
This is also just about everywhere in the code.

jerrypeng replied:
will change

revans2 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCenter.java:
```diff
@@ -0,0 +1,238 @@
+            }
+            return new ArrayList<Hierarchy>(hierarchies.values());
+        } catch (Exception e) {
+            LOG.error("Get hierarchies error", e);
```
I realize a lot of this code came directly from JStorm, but I really would prefer to not eat exceptions.  If CGroups is supposed to be enabled I would rather have the supervisor crash if there is an issue running instead of disabling it and outputting an error message.

revans2 replied:
And this is just about everywhere in the code.

jerrypeng replied:
perhaps we can file a follow up jira to investigate which functions we want to throw exceptions vs. eating them

revans2 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCenter.java:
```diff
@@ -0,0 +1,238 @@
+
+        Set<SubSystem> subSystems = this.getSubSystems();
+        for (SubSystem subSystem : subSystems) {
+            if (subSystem.getType() == subsystem)
```
Can you please update all of the blocks to include `{` and `}` even for single line blocks.

jerrypeng replied:
will fix

revans2 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCenter.java:
```diff
@@ -0,0 +1,238 @@
+    public void mount(Hierarchy hierarchy) throws IOException {
+
+        if (this.mounted(hierarchy) != null) {
+            LOG.error(hierarchy.getDir() + " is mounted");
```
Can we please update the log messages to use "{}" substitution.

jerrypeng replied:
will fix

revans2 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCommonOperation.java:
```diff
@@ -0,0 +1,47 @@
+
+    public void setEventControl(String eventFd, String controlFd, String... args) throws IOException;
+
+}
```
IS there any way to add some javadocs to these.  It would make it a lot simpler to understand what each of these are supposed to do.

jerrypeng replied:
will add

revans2 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupUtils.java:
```diff
@@ -0,0 +1,174 @@
+            String cmd = "rmdir " + dir;
+            SystemOperation.exec(cmd);
+        } catch (IOException e) {
+            // TODO Auto-generated catch block
```
Please remove this TODO.

jerrypeng replied:
will fix

revans2 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupUtils.java:
```diff
@@ -0,0 +1,174 @@
+    public static void deleteDir(String dir) {
+        try {
+            String cmd = "rmdir " + dir;
+            SystemOperation.exec(cmd);
```
I know that this came from JStorm, but is there a reason we should not just do 
```
File d = new File(dir);
if (d.exists()) {
    if (d.isDirectory()) {
        d.delete(); //TODO need to check results
    } else {
        throw some error;
    }
} 
```

The current code is not safe, it opens up the potential for command injection.

revans2 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupUtils.java:
```diff
@@ -0,0 +1,174 @@
+    }
+
+    public static void writeFileByLine(String fileDir, String string) throws IOException {
+        LOG.info("For CGroups - writing {} to {} ", string, fileDir);
```
Can we make this debug?

jerrypeng replied:
will fix

revans2 commented on storm-core/test/jvm/org/apache/storm/TestCGroup.java:
```diff
@@ -0,0 +1,94 @@
+package org.apache.storm;
```
Needs Apache Header

jerrypeng replied:
will add

jerrypeng replied:
added

revans2 commented on storm-core/test/jvm/org/apache/storm/TestCGroup.java:
```diff
@@ -0,0 +1,94 @@
+/**
+ * Created by jerrypeng on 1/19/16.
+ */
+public class TestCGroup {
```
Needs real javadocs or remove the comment.

jerrypeng replied:
will fix and add

jerrypeng commented:
@revans2 thank you for the review.  I added the apache headers to files that were missing them and addressed all of your comments

bastiliu commented on storm-core/src/clj/org/apache/storm/daemon/supervisor.clj:
```diff
@@ -1088,8 +1104,26 @@
                                           storm-log4j2-conf-dir)
                                      file-path-separator "worker.xml")
+
+          cgroup-command (if (conf STORM-CGROUP-ENABLE)
```
Maybe it is better to disable cgroup in local mode since the cgroup is not supported in Windows or some linux with old core version

jerrypeng replied:
cgroups is disabled by default (via default.yaml) and the code you referenced is the code that will be executed for "Cluster mode" and not "local mode".

vesense commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCenter.java:
```diff
@@ -0,0 +1,238 @@
+
+public class CgroupCenter implements CgroupOperation {
+
+    public static Logger LOG = LoggerFactory.getLogger(CgroupCenter.class);
```
public -> private

jerrypeng replied:
will fix

vesense commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCommon.java:
```diff
@@ -0,0 +1,225 @@
+
+    private final Set<CgroupCommon> children = new HashSet<CgroupCommon>();
+
+    public static final Logger LOG = LoggerFactory.getLogger(CgroupCommon.class);
```
public -> private

jerrypeng replied:
will fix

vesense commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupManager.java:
```diff
@@ -0,0 +1,176 @@
+
+public class CgroupManager implements ResourceIsolationInterface {
+
+    public static final Logger LOG = LoggerFactory.getLogger(CgroupManager.class);
```
public -> private

jerrypeng replied:
will fix

vesense commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupUtils.java:
```diff
@@ -0,0 +1,133 @@
+
+public class CgroupUtils {
+
+    public static final Logger LOG = LoggerFactory.getLogger(CgroupUtils.class);
```
public -> private

jerrypeng replied:
will fix

vesense commented on storm-core/src/jvm/org/apache/storm/container/cgroup/SystemOperation.java:
```diff
@@ -0,0 +1,70 @@
+
+public class SystemOperation {
+
+    public static final Logger LOG = LoggerFactory.getLogger(SystemOperation.class);
```
public -> private

jerrypeng replied:
will fix

vesense commented on storm-core/src/jvm/org/apache/storm/container/cgroup/SystemOperation.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    public static void main(String[] args) throws IOException {
+        SystemOperation.mount("test", "/cgroup/cpu", "cgroup", "cpu");
```
Maybe we can add some unit Tests for CGroups and remove the `main` method code.

jerrypeng replied:
will remove main method, but perhaps we can file a follow up jira to cover all these helper functions since this PR is pretty big as it is and I have a include a unit test that tests the end to end functionality of CGroups.  A lot of the CGroups code came from JStorm, but I was unable to find any unit tests for them in JStorm.  Perhaps some that worked on JStorm can shed some light on whether there are individual unit tests for these functions

vesense commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCenter.java:
```diff
@@ -0,0 +1,238 @@
+    }
+
+    public static void main(String args[]) {
+        System.out.println(CgroupCenter.getInstance().getHierarchies().get(0).getRootCgroups().getChildren().size());
```
Maybe we can add some unit Tests for CGroups and remove the `main` method code.

jerrypeng replied:
will remove the main method, and please refer to my previous comment

vesense commented on storm-core/src/jvm/org/apache/storm/utils/Utils.java:
```diff
@@ -1382,5 +1385,60 @@ public static TopologyInfo getTopologyInfo(String name, String asUser, Map storm
+        try {
+            Thread.sleep(ms);
+        } catch (InterruptedException e) {
+
```
Do we need: `throw new RuntimeException(e);`

jerrypeng replied:
will add

bastiliu commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupManager.java:
```diff
@@ -0,0 +1,176 @@
+        if (cpuNum != null) {
+            CpuCore cpuCore = (CpuCore) workerGroup.getCores().get(SubSystemType.cpu);
+            try {
+                cpuCore.setCpuShares(cpuNum.intValue());
```
cpu.share might be not a good choice. We had a testing before. The real usage of cpu for a process can not be limited as expected in some scenarios. e.g. there are two processes which have same cpu weight configured, and the node has 2 cores. It is expected that each process occupy 1 core. If the load of these processes are similar (occupy similar cpu cores), the control of final usage for each process  is good by cpu.share. Then, if one process become idle, the other process can occupy all cpu cores. It will not be consistent with the configuration in RAS.

jerrypeng replied:
cpu.share config limits cpu usage based on weights and is only relative to what else is running on the node.  Though the cpu.share can serve as a guarantee of resources not necessary as a hard limit.  In the scenario you described, why not let the one process use more cpu resources if those cpu resources are free?  Only when there is a lot of contention of cpu resources do we need to make sure everyone is given the amount of cpu as per their guarantee, but if the node is pretty free, why not let a worker use more resources than technically is guaranteed/allocated to that worker.  That is kind of why I think.

jerrypeng replied:
The downside to this is that a topology might perform different when the cluster is relatively free vs if the cluster is heavily utilized.  This may give a user a harder time to judge how much resources to allocate to a topology to achieve a certain performance.  However, if we use cpu.share to limit cpu, clusters may achieve better cpu utilization.  And using cpu.share will also give storm clusters some inherent elasticity, since a topology's use of CPU resources can grow (as long as there are free resources) and shrink depending on need since there are no hard limits set.  By using cpu.shares we can have this elastic benefit and at the same time have a way to guarantee a certain amount of resources.

revans2 replied:
Would it be worth filing a follow on JIRA so that we can control if it is a hard CPU limit or a CPU guarantee?

jerrypeng replied:
We can file a follow on JIRA to add a hard CPU limit on a per worker basis.  Currently, the code does support setting a upper hard limit to how many cores/cpu resources all of the workers on a node can use (e.g. if a node has 24 cores but you want workers to use only 12 of the cores, you can).

bastiliu replied:
I am OK for filing follow-up JIRA of hard CPU limit or CPU guarantee. But we still need a full testing on cpu.share, whether the worker which come back from idle to busy can retrieve enough cpu resource again instead of starving, when the other worker has occupied all cpu resources. Actually, we has met this kind problem on online jobs. That is reason why we choose a hard limit finally.

jerrypeng replied:
ok, I agree we should test this feature to make sure idle workers can re-acquire the resources they are guaranteed once they "wake up".

jerrypeng replied:
Anyways adding cpu.cfs_period_us and cpu.cfs_quota_us to set a hard time should be a minor and easy change to the current code in the PR, if we find that cpu.share doesn't work well

jerrypeng commented:
@revans2 @bastiliu @vesense Thanks for the reviews! I really appreciated it!  I think I have addressed all of your comments.  Does anyone have any other concerns?

redsanket commented on storm-core/test/jvm/org/apache/storm/TestCGroup.java:
```diff
@@ -0,0 +1,118 @@
+/**
```
Can we have the filenames also to be consistent, TestCgroup.java. Not a big deal though

jerrypeng replied:
will fix

redsanket commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCommon.java:
```diff
@@ -0,0 +1,225 @@
+    }
+
+    private void free() throws IOException {
+        for (CgroupCommon child : this.children)
```
no paranthesis

jerrypeng replied:
will add

redsanket commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupManager.java:
```diff
@@ -0,0 +1,176 @@
+
+    private CgroupCenter center;
+
+    private Hierarchy h;
```
h does not seems to be a valid name, can we have something like hierarchy

jerrypeng replied:
will rename

redsanket commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupManager.java:
```diff
@@ -0,0 +1,176 @@
+    }
+
+    /**
+     * User cfs_period & cfs_quota to control the upper limit use of cpu core e.g. If making a process to fully use two cpu cores, set cfs_period_us to
```
Can we take split the comment to next line for readability

jerrypeng replied:
will spit

redsanket commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupManager.java:
```diff
@@ -0,0 +1,176 @@
+    }
+
+    public String startNewWorker(String workerId, Map resourcesMap) throws SecurityException {
+        Number cpuNum = (Number) resourcesMap.get("cpu");
```
why are we using a Number, are we not sure of its type

jerrypeng replied:
since users can submit a value representing the amount of resources as an integer or double from RAS

redsanket commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupManager.java:
```diff
@@ -0,0 +1,176 @@
+            totalMem = (Number) resourcesMap.get("memory");
+        }
+
+        CgroupCommon workerGroup = new CgroupCommon(workerId, h, this.rootCgroup);
```
h here is hard to read what it is exactly

jerrypeng replied:
renamed

redsanket commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupManager.java:
```diff
@@ -0,0 +1,176 @@
+
+        sb.append(this.conf.get(Config.STORM_CGROUP_CGEXEC_CMD)).append(" -g ");
+
+        Iterator<SubSystemType> it = h.getSubSystems().iterator();
```
same here h?

jerrypeng replied:
renamed

redsanket commented on storm-core/src/jvm/org/apache/storm/container/cgroup/core/MemoryCore.java:
```diff
@@ -0,0 +1,189 @@
+            this.totalActiveFileSize = Long.parseLong(splits[22]);
+            this.totalUnevictableSize = Long.parseLong(splits[23]);
+            this.totalHierarchicalMemoryLimitSize = Long.parseLong(splits[24]);
+            this.totalHierarchicalMemswLimitSize = Long.parseLong(splits[25]);
```
I presume replacing it with totalHierarchialMemSwapLimitSize would be more readable, if it is what it means

redsanket replied:
Just a few minor comments, the core implementation of cgroup manager seems fine, not sure how other subsytems come into the picture if we are using only cpu and memory. @revans2 might want to take a look at it before it gets merged.

redsanket commented on storm-core/test/jvm/org/apache/storm/TestCGroup.java:
```diff
@@ -0,0 +1,118 @@
+        byte[] data = Files.readAllBytes(Paths.get(filePath));
+        return new String(data).trim();
+    }
+}
```
new line at the end of file

jerrypeng replied:
will add

jerrypeng commented:
@redsanket Thanks for your review.  I addressed all your comments

knusbaum commented on conf/cgconfig.conf.example:
```diff
@@ -0,0 +1,41 @@
+       }
+       cpu {
+       }
+}
```
Newline.

jerrypeng replied:
will fix

knusbaum commented on storm-core/src/clj/org/apache/storm/daemon/supervisor.clj:
```diff
@@ -344,6 +348,12 @@
    :download-lock (Object.)
    :stormid->profiler-actions (atom {})
+   :cgroup-manager (if (conf STORM-CGROUP-ENABLE)
+                     (let [cgroup-manager (.newInstance (Class/forName (conf STORM-RESOURCE-ISOLATION-PLUGIN)))]
```
If the resource isolation plugin is pluggable, do we want to call this `:cgroup-manager`?

jerrypeng replied:
will refactor

knusbaum commented on storm-core/src/clj/org/apache/storm/daemon/supervisor.clj:
```diff
@@ -344,6 +348,12 @@
    :sync-retry (atom 0)
    :download-lock (Object.)
    :stormid->profiler-actions (atom {})
+   :cgroup-manager (if (conf STORM-CGROUP-ENABLE)
```
Also, why do we only load the isolation plugin if cgroups is enabled? Maybe this config can go away and we can just check if `storm.resource.isolation.plugin` is set?

knusbaum commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCenter.java:
```diff
@@ -0,0 +1,232 @@
+    }
+
+    /**
+     * Thread unsafe
```
This comment is useless. Please clarify or remove

jerrypeng replied:
will remove

knusbaum commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCenter.java:
```diff
@@ -0,0 +1,232 @@
+                String name = strSplit[0];
+                String type = strSplit[3];
+                String dir = strSplit[1];
+                Hierarchy h = hierarchies.get(type);
```
Why bother assigning here?

jerrypeng replied:
will remove

knusbaum commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCenter.java:
```diff
@@ -0,0 +1,232 @@
+                    continue;
+                }
+                subSystems.add(new SubSystem(type, Integer.valueOf(split[1]), Integer.valueOf(split[2])
+                        , Integer.valueOf(split[3]).intValue() == 1 ? true : false));
```
Move comma next to previous argument.

jerrypeng replied:
will refactor

d2r replied:
` ? true : false` can be deleted

knusbaum commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCenter.java:
```diff
@@ -0,0 +1,232 @@
+    public boolean enabled(SubSystemType subsystem) {
+
+        Set<SubSystem> subSystems = this.getSubSystems();
+        for (SubSystem subSystem : subSystems) {
```
`subSystem` vs `subsystem` is confusing.

jerrypeng replied:
will refactor

knusbaum commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupUtils.java:
```diff
@@ -0,0 +1,133 @@
+        }
+    }
+
+    public static void writeFileByLine(String fileDir, String string) throws IOException {
```
This whole method's body could be
```
{
    writeFileByLine(fileDir, Arrays.asList(string));
}
```

jerrypeng replied:
will refactor

kishorvpatil commented on storm-core/src/clj/org/apache/storm/daemon/supervisor.clj:
```diff
@@ -301,7 +303,9 @@
-    (try-cleanup-worker conf id))
+    (try-cleanup-worker conf id)
+    (if (conf STORM-CGROUP-ENABLE)
+      (.shutDownWorker (:cgroup-manager supervisor) id false)))
```
I would move check and call for `shutdownWorker` under `try-cleanup-worker` method itself. Also, we can check if `(not-nil? (:cgroup-manger supervisor) )`.

jerrypeng replied:
will move to try-clean-up worker.  I don't think we need at to check for null since we set a default value for storm.resource.isolation.plugin config.

knusbaum commented on storm-core/src/jvm/org/apache/storm/container/cgroup/Constants.java:
```diff
@@ -0,0 +1,30 @@
+ */
+package org.apache.storm.container.cgroup;
+
+public class Constants {
```
I don't think we probably want a class called `Constants`.

jerrypeng replied:
will move out

knusbaum commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupUtils.java:
```diff
@@ -0,0 +1,133 @@
+        return result;
+    }
+
+    public static String reAnalyse(Set<SubSystemType> subSystems) {
```
Why is this called `reAnalyse`? What does that mean?

d2r replied:
I agree this is confusing. `analyse` seems to parse the CSV string into a Set of SubSystemTypes, and `reAnalyse` seems to do the reverse. It is not necessarily true that `reAnalyse(analyse(o)).equals(o)`.

jerrypeng replied:
will rename and add comment

knusbaum commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupUtils.java:
```diff
@@ -0,0 +1,133 @@
+        }
+    }
+
+    public static boolean fileExists(String dir) {
```
Don't we have Utils functions for this type of thing?

jerrypeng replied:
will refactor

knusbaum commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupUtils.java:
```diff
@@ -0,0 +1,133 @@
+        return file.exists();
+    }
+
+    public static boolean dirExists(String dir) {
```
And this?

jerrypeng replied:
will move to util.Utils

knusbaum commented on storm-core/src/jvm/org/apache/storm/container/cgroup/Hierarchy.java:
```diff
@@ -0,0 +1,117 @@
+        if (getClass() != obj.getClass()) {
+            return false;
+        }
+        Hierarchy other = (Hierarchy) obj;
```
Could we just do other.hashCode() == this.hashCode() at this point?

knusbaum commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupOperation.java:
```diff
@@ -0,0 +1,46 @@
+import java.util.List;
+import java.util.Set;
+
+public interface CgroupOperation {
```
Can we get some javadocs on this interface? It's not at all obvious what many of these methods are intended to do.

jerrypeng replied:
will refactor and add comments

knusbaum commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCenter.java:
```diff
@@ -0,0 +1,232 @@
+                    ret = null;
+                    break;
+                }
+            }
```
So if the last Hierarchy element in heirarchies has a set of subsystems that doesn't contain all of `subSystems`, then this 'center' is busy?

Why loop through all the hierarchies?

knusbaum commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCenter.java:
```diff
@@ -0,0 +1,232 @@
+    }
+
+    @Override
+    public Hierarchy busy(SubSystemType subsystem) {
```
```
{
    return busy(Arrays.toList(subsystem));
}
```

jerrypeng replied:
will refactor

knusbaum commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCommon.java:
```diff
@@ -0,0 +1,226 @@
+        CgroupUtils.deleteDir(this.dir);
+    }
+
+    private void init() {
```
Should we have `init()` up by the constructors where it's called? Seems silly to have it at the bottom.

jerrypeng replied:
will refactor

knusbaum commented on storm-core/src/jvm/org/apache/storm/container/cgroup/Device.java:
```diff
@@ -0,0 +1,72 @@
+        if (getClass() != obj.getClass()) {
+            return false;
+        }
+        Device other = (Device) obj;
```
`.hashcode() == .hashcode()`?

knusbaum commented on storm-core/src/jvm/org/apache/storm/container/cgroup/SubSystem.java:
```diff
@@ -0,0 +1,78 @@
+    public boolean equals(Object object) {
+        boolean ret = false;
+        if (object != null && object instanceof SubSystem) {
+            ret = (this.type.equals(((SubSystem)object).getType()) && this.hierarchyID == ((SubSystem)object).getHierarchyID());
```
Cant enums be compared with `==` as well?

jerrypeng replied:
will refactor

knusbaum commented on storm-core/src/jvm/org/apache/storm/container/cgroup/SystemOperation.java:
```diff
@@ -0,0 +1,65 @@
+
+    public static void mount(String name, String target, String type, String data) throws IOException {
+        StringBuilder sb = new StringBuilder();
+        sb.append("mount -t ").append(type).append(" -o ").append(data).append(" ").append(name).append(" ").append(target);
```
Can we break this up into lines? I think it would actually be easier to tell what's being appended if the things being appended are all aligned vertically.

```
sb.append("mount -t ")
    .append(type)
    .append(" -o ")
    .append(data)
    .append(" ")
    .append(name)
    .append(" ")
    .append(target);
```

jerrypeng replied:
will refactor

knusbaum commented on storm-core/src/jvm/org/apache/storm/container/cgroup/SystemOperation.java:
```diff
@@ -0,0 +1,65 @@
+            throw new IOException(ie.toString());
+        }
+    }
+}
```
newline

jerrypeng replied:
will add

knusbaum commented on storm-core/src/jvm/org/apache/storm/container/cgroup/core/CpuCore.java:
```diff
@@ -0,0 +1,136 @@
+            if (getClass() != obj.getClass()) {
+                return false;
+            }
+            Stat other = (Stat) obj;
```
`hashCode() == hashCode()`?

knusbaum commented on storm-core/src/jvm/org/apache/storm/container/cgroup/core/BlkioCore.java:
```diff
@@ -0,0 +1,259 @@
+            result.put(device, iops);
+        }
+        return result;
+    }
```
`getBlkioWeightDevice`, `getReadBps`, `getWriteBps`, `getReadIOps`, `getWriteIOps`

Don't Repeat Yourself. Same with the set versions.

knusbaum commented on storm-core/src/jvm/org/apache/storm/container/cgroup/core/BlkioCore.java:
```diff
@@ -0,0 +1,259 @@
+        return result;
+    }
+
+    public Map<Device, Long> getBlkioSectors() throws IOException {
```
More D.R.Y.

knusbaum commented on storm-core/src/jvm/org/apache/storm/container/cgroup/core/CpusetCore.java:
```diff
@@ -0,0 +1,212 @@
+        return parseNums(output);
+    }
+
+    public void setMems(int[] nums) throws IOException {
```
D.R.Y. `setCpus`

knusbaum commented on storm-core/test/clj/org/apache/storm/supervisor_test.clj:
```diff
@@ -335,8 +334,8 @@
-                                      mock-mem-onheap)
-                (verify-first-call-args-for-indices launch-process
+                                      (WorkerResources.))
+            (verify-first-call-args-for-indices launch-process
```
Wrong indentation.

jerrypeng replied:
will fix

knusbaum commented on storm-core/test/clj/org/apache/storm/supervisor_test.clj:
```diff
@@ -362,8 +361,8 @@
-                                      mock-mem-onheap)
-                (verify-first-call-args-for-indices launch-process
+                                      (WorkerResources.))
+            (verify-first-call-args-for-indices launch-process
```
Wrong indentation.

jerrypeng replied:
will fix

knusbaum commented:
Done with first pass. This code could do with a lot more documentation. 
I don't want to merge a massive chunk of undocumented code.

d2r commented on conf/defaults.yaml:
```diff
@@ -281,3 +281,16 @@ pacemaker.thread.timeout: 10
+    - cpu
+    - memory
+storm.cgroup.hierarchy.name: "storm"
+# Also determines whether the unit tests for cgroup runs.  If cgroup.enable is set to false the unit tests for cgroups will not run
```
I think it would be better if this comment were part of the javadoc for the config definition in Config.java.  If we remove `storm.cgroup.enable` as @knusbaum [suggests](https://github.com/apache/storm/pull/1053/files#r52045815), then this comment should move to the plugin config too.

jerrypeng replied:
will refactor

d2r commented on storm-core/src/clj/org/apache/storm/daemon/supervisor.clj:
```diff
@@ -42,7 +42,9 @@
   (:gen-class
-    :methods [^{:static true} [launch [org.apache.storm.scheduler.ISupervisor] void]]))
+    :methods [^{:static true} [launch [org.apache.storm.scheduler.ISupervisor] void]])
+  (:import [org.apache.storm.container.cgroup CgroupManager])
```
CgroupManager unused?

jerrypeng replied:
removed

d2r commented on storm-core/src/jvm/org/apache/storm/container/ResourceIsolationInterface.java:
```diff
@@ -0,0 +1,43 @@
+     * @param resources set of resources to limit
+     * @return a String that includes to command on how to start the worker.  The string returned from this function
+     * will be concatenated to the front of the command to launch logwriter/worker in supervisor.clj
+     */
```
minor: `The string returned`... Maybe we should describe that the string is a series of command line options for launching a JVM. It seems strange to document what the caller does with the string.

d2r commented on storm-core/src/jvm/org/apache/storm/container/ResourceIsolationInterface.java:
```diff
@@ -0,0 +1,43 @@
+     * @return a String that includes to command on how to start the worker.  The string returned from this function
+     * will be concatenated to the front of the command to launch logwriter/worker in supervisor.clj
+     */
+    public String startNewWorker(String workerId, Map resources);
```
minor: these method declarations do not need `public`

jerrypeng replied:
will remove

d2r commented on storm-core/src/clj/org/apache/storm/daemon/supervisor.clj:
```diff
@@ -381,7 +391,7 @@
                 port
                 id
-                mem-onheap)
+                resources)
```
Let's remove the binding for `mem-onheap` above since we are no longer using it.

jerrypeng replied:
will remove

d2r commented on storm-core/src/clj/org/apache/storm/daemon/supervisor.clj:
```diff
@@ -1062,9 +1072,15 @@
+          mem-onheap (if (and (.get_mem_on_heap resources) (> (.get_mem_on_heap resources) 0)) ;; not nil and not zero
+                       (int (Math/ceil (.get_mem_on_heap resources))) ;; round up
                        (storm-conf WORKER-HEAP-MEMORY-MB)) ;; otherwise use default value
+          mem-offheap (if (.get_mem_off_heap resources)
```
Non-negative check for off-heap?

jerrypeng replied:
will add

jerrypeng replied:
we actually don't need to check any of the resource configs since the configs are validated in StormSubmitter when the topology is submitted

d2r commented on storm-core/src/clj/org/apache/storm/daemon/supervisor.clj:
```diff
@@ -1062,9 +1072,15 @@
+                        (int (Math/ceil (.get_mem_off_heap resources))) ;; round up
+                        0)
+
+          cpu (int (Math/ceil (.get_cpu resources)))
```
Is the "cpu" resource already validated, or do we need to check its values here?

jerrypeng replied:
we actually don't need to check any of the resource configs since the configs are validated in StormSubmitter when the topology is submitted

d2r commented on storm-core/src/clj/org/apache/storm/daemon/supervisor.clj:
```diff
@@ -1088,8 +1104,26 @@
+                           (str/split
+                             (.startNewWorker (:cgroup-manager supervisor) worker-id
+                               (merge
+                                 ;; The manually set CGROUP-WORKER-CPU-LIMIT config on supervisor will overwrite resources assigned by RAS (Resource Aware Scheduler)
```
Check comment; seems it should refer to memory.

jerrypeng replied:
will fix

d2r commented on storm-core/src/clj/org/apache/storm/daemon/supervisor.clj:
```diff
@@ -1178,7 +1212,7 @@
 
 (defmethod launch-worker
-    :local [supervisor storm-id port worker-id mem-onheap]
+    :local [supervisor storm-id port worker-id resources]
```
revert?

jerrypeng replied:
why revert? Shouldn't we try the keep all launch-worker definitions consistent?

d2r replied:
I think when I made my comment I thought we had simply added an unused parameter, but that was not the case. Yes, we should keep things consistent as you have done, so disregard.

d2r commented on storm-core/src/jvm/org/apache/storm/Config.java:
```diff
@@ -2155,6 +2156,63 @@
 
+
+    @isImplementationOfClass(implementsClass = ResourceIsolationInterface.class)
+    public static final Object STORM_RESOURCE_ISOLATION_PLUGIN = "storm.resource.isolation.plugin";
```
It would be helpful to have a javadoc comment for this config.

jerrypeng replied:
will add

d2r commented on conf/defaults.yaml:
```diff
@@ -281,3 +281,16 @@ pacemaker.thread.timeout: 10
+storm.cgroup.hierarchy.dir: "/cgroup/storm_resources"
+storm.cgroup.resources:
+    - cpu
+    - memory
```
With few exceptions, I see string values are quoted. Not sure this would make a difference with these values, but maybe we should be consistent.

jerrypeng replied:
will add quotes

d2r commented on storm-core/src/jvm/org/apache/storm/container/cgroup/SubSystemType.java:
```diff
@@ -0,0 +1,58 @@
+        }
+        else if (str.equals("net_prio")) {
+            return net_prio;
+        }
```
Can we rewrite this as a `switch` ?

jerrypeng replied:
will refactor

d2r commented on storm-core/src/jvm/org/apache/storm/container/cgroup/SubSystemType.java:
```diff
@@ -0,0 +1,58 @@
+
+public enum SubSystemType {
+
+    // net_cls,ns is not supposted in ubuntu
```
Do we need this comment?

jerrypeng replied:
will remove

d2r commented on storm-core/src/jvm/org/apache/storm/container/cgroup/Hierarchy.java:
```diff
@@ -0,0 +1,117 @@
+        result = prime * result + ((name == null) ? 0 : name.hashCode());
+        result = prime * result + ((type == null) ? 0 : type.hashCode());
+        return result;
+    }
```
Are two instances of Hierarchy considered identical even when `subSystems` and `type` are different?

jerrypeng replied:
the member variable "type" is just a comma delimited string of subsystems that the hierarchy uses.  So in terms of identification, "type" and "subSystems" are the same

d2r replied:
I see, so since one is generated from the other, we do not need to check both.

kishorvpatil commented:
The proposed changes are launching commands using `exec.command` which may not work in multi-tenant setup ( for killing workers etc.). It will not work in multi-tenant mode.

kishorvpatil commented:
Secondly, I don't necessarily like the idea of `ResourceIsolationInterface` taking up responsibility for starting and stopping workers.. This would be moving a lot of responsibility from supervisor to `ResourceIsolationInterface` specially when we do a lot of setting-up/tearing-down for worker and use worker-launcher in multitenant cases. I would have simple interface which may help us call from within supervisor - and possibly more than once..- without any side-effects so that we may call them again in case of partial failures.
```
public interface ResourceAllocationInterface {
boolean reserveResourcesForWorker(workerid, resources);
boolean releaseResourcesForWorker(workerid);
```

knusbaum commented on storm-core/src/jvm/org/apache/storm/utils/Utils.java:
```diff
@@ -1367,12 +1367,67 @@ public static int toPositive(int number) {
+        }
+    }
+
+    public static void kill(Integer pid) {
```
We already have process killing stuff in Util.java.

jerrypeng replied:
get rid of

knusbaum commented on storm-core/src/jvm/org/apache/storm/utils/Utils.java:
```diff
@@ -1367,12 +1367,67 @@ public static int toPositive(int number) {
+        exec.execute(cmd);
+    }
+
+    public static void sleepMs(long ms) {
```
And sleep stuff in `Util` or `Time` or something.

jerrypeng replied:
get rid of

d2r commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCenter.java:
```diff
@@ -0,0 +1,232 @@
+            }
+        }
+        return false;
+    }
```
I would like to rename the parameter (it is a type, not a subsystem), and it would be good to rename the method to something like `isSubsystemTypeEnabled`.

jerrypeng replied:
will rename isSubsystemEnabled

d2r commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCenter.java:
```diff
@@ -0,0 +1,232 @@
+            }
+        }
+        return null;
+    }
```
It would be good to rename `busy` and perhaps add a javadoc explaining what it does.

jerrypeng replied:
will rename

jerrypeng commented:
@d2r @knusbaum @kishorvpatil Thank you for your reviews.  I have addressed all your comments.

I have also added additional documentation in the code.

@kishorvpatil I rewrote the ResourceIsolationInterface as per our discussion

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCenter.java:
```diff
@@ -0,0 +1,222 @@
+        if (instance == null) {
+            instance = new CgroupCenter();
+        }
+        return CgroupUtils.enabled() ? instance : null;
```
should this be the first line instead?

jerrypeng replied:
will change

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCenter.java:
```diff
@@ -0,0 +1,222 @@
+             BufferedReader br = new BufferedReader(reader)) {
+            String str = null;
+            while ((str = br.readLine()) != null) {
+                String[] strSplit = str.split(" ");
```
should there be a check on length of strSplit here?

jerrypeng replied:
I don't think its necessary.  The format for the cgroup files are set and the whole statement is in a try statement to catch any exception.

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCenter.java:
```diff
@@ -0,0 +1,222 @@
+                String type = strSplit[3];
+                String dir = strSplit[1];
+                Hierarchy h = new Hierarchy(name, CgroupUtils.getSubSystemsFromString(type), dir);
+                hierarchies.put(type, h);
```
is the map being used to generate unique entries for each type? If so, do you want to log a warn message for two entries with same type

jerrypeng replied:
"type" is just a comma delimited list of subsystems that are mounted on a hierarchy. Cgroup does not support mounting the same set of subsystems on multiple heirarchies (a.k.a mirrored hierarchies).  Thus, using the comma delimited list of subsystems as key is going to be distinct

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCenter.java:
```diff
@@ -0,0 +1,222 @@
+             BufferedReader br = new BufferedReader(reader)){
+            String str = null;
+            while ((str = br.readLine()) != null) {
+                String[] split = str.split("\t");
```
again length check on split?

jerrypeng replied:
Not necessary please see my previous comment on this subject manner.

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCenter.java:
```diff
@@ -0,0 +1,222 @@
+                if (type == null) {
+                    continue;
+                }
+                subSystems.add(new SubSystem(type, Integer.valueOf(split[1]), Integer.valueOf(split[2]),
```
Instead of using split[1], split[2] directly, it will be good to have declared variables e.g. name = split[1]. Similar to what you have done in getHierarchies.

jerrypeng replied:
will refactor

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCenter.java:
```diff
@@ -0,0 +1,222 @@
+
+        Set<SubSystem> subSystems = this.getSubSystems();
+        for (SubSystem subSystem : subSystems) {
+            if (subSystem.getType() == subSystemType) {
```
It should use equals instead of identity check

jerrypeng replied:
for comparing enums .equals or "==" is the same

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCenter.java:
```diff
@@ -0,0 +1,222 @@
+
+    @Override
+    public boolean isMounted(Hierarchy hierarchy) {
+
```
There is new line such as this in other functions as well. It can be removed.

jerrypeng replied:
will remove

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCenter.java:
```diff
@@ -0,0 +1,222 @@
+    public boolean isMounted(Hierarchy hierarchy) {
+
+        List<Hierarchy> hierarchies = this.getHierarchies();
+        if (Utils.CheckDirExists(hierarchy.getDir())) {
```
This condition should be before this.getHierarchies.

abhishekagarwal87 replied:
Also, getHierarchies is going and reading from the file system. If getHierarchies is being called too often, you should consider caching.

jerrypeng replied:
will move this.getHierarchies() under the if conditional

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCenter.java:
```diff
@@ -0,0 +1,222 @@
+        for (SubSystemType type : subSystems) {
+            if (this.getHierarchyWithSubSystem(type) != null) {
+                LOG.error("subSystem: {} is busy", type.name());
+                subSystems.remove(type);
```
It modifies the contents of hierarchy object also which can have side effects. Do you want to duplicate subSystems?

jerrypeng replied:
this is the wanted effect this function.  We only want to mount the subsystems that are not busy or can be mounted.  A Hierarchy object is manually instantiated by a user and then the mount function can be called to actually create the hierarchy and mount the subsystems.

abhishekagarwal87 replied:
Understood now.

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCenter.java:
```diff
@@ -0,0 +1,222 @@
+        while (parent != null) {
+            if (!Utils.CheckDirExists(parent.getDir())) {
+                LOG.error("Parent {} does not exist", parent.getDir());
+                return;
```
should it rather throw some exception or return a boolean result?

jerrypeng replied:
I think printing out an error message is ok for now.  Since we are not sure if we want to crash the supervisor if something in cgroups goes wrong.  Though this is something to investigate in a follow jira

jerrypeng replied:
actually this is a important function, lets throw an exception

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCenter.java:
```diff
@@ -0,0 +1,222 @@
+            return;
+        }
+
+        //Todo perhaps thrown exception or print out error message is dir is not created successfully
```
this todo can be removed.

jerrypeng replied:
will remove

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCommon.java:
```diff
@@ -0,0 +1,247 @@
+    public static final String RELEASE_AGENT = "/release_agent";
+    public static final String CGROUP_CLONE_CHILDREN = "/cgroup.clone_children";
+    public static final String CGROUP_EVENT_CONTROL = "/cgroup.event_control";
+    public static final String CGROUP_PROCS = "/cgroup.procs";
```
would be better to use a consistent patter e.g. cgroup_procs

jerrypeng replied:
these are not just arbitrary strings but actual names for files used in cgroup

abhishekagarwal87 replied:
My bad. I definitely need to read up more on cgroups to understand it better.

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCommon.java:
```diff
@@ -0,0 +1,247 @@
+
+    private final CgroupCommon parent;
+
+    //private final Map<SubSystemType, CgroupCore> cores;
```
remove the commented code

jerrypeng replied:
will remove

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCommon.java:
```diff
@@ -0,0 +1,247 @@
+     * rootCgroup
+     */
+    public CgroupCommon(Hierarchy hierarchy, String dir) {
+        this.name = "";
```
should the name be same as dir?

jerrypeng replied:
the name variable holds the path relative to storm base cgroup directory while dir holds the complete path

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCommon.java:
```diff
@@ -0,0 +1,247 @@
+
+    @Override
+    public boolean getNotifyOnRelease() throws IOException {
+        return CgroupUtils.readFileByLine(CgroupUtils.getDir(this.dir, NOTIFY_ON_RELEASE)).get(0).equals("1") ? true : false;
```
can you put a check that number of returned rows by readFileByLine is more than zero.

jerrypeng replied:
its not necessary since its assumed that there will always be a value in that file.

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupUtils.java:
```diff
@@ -0,0 +1,128 @@
+        return result;
+    }
+
+    public static void writeFileByLine(String fileDir, List<String> strings) throws IOException {
```
fileDir can be renamed to filepath. strings can be renames to lines.

jerrypeng replied:
will refactor

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCommon.java:
```diff
@@ -0,0 +1,247 @@
+    }
+
+    @Override
+    public void setNotifyOnRelease(boolean flag) throws IOException {
```
what is the expected behavior if this is called again? Currently it doesn't override the existing flag but appends it

jerrypeng replied:
It doesn't matter.  All the cgroup config "files" only takes the the latest value written to the file

abhishekagarwal87 replied:
I understand that. But cgroup config files should one value only in the file at a time? writeFileByLine should override the file with latest value. If you read getNotifyOnRelease, you will read the first line of file which is actually the oldest. please refer to my comment on writeFileByLine

jerrypeng replied:
The cgroup config file is not actually mounted to a filesystem thus a write to it is not written to disk necessary.  The config files serve as a way to pass in parameters into cgroups like linux kernel modules sometimes do.  For example:

$ cat cpu.shares
75
$ echo "80" >> cpu.shares
$ cat cpu.shares
80
$ echo "75" > cpu.shares
$ cat cpu.shares
75

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCommon.java:
```diff
@@ -0,0 +1,247 @@
+    @Override
+    public void setReleaseAgent(String command) throws IOException {
+        if (!this.isRoot) {
+            return;
```
should you log an error/warn if it is called from non-root cgroup

jerrypeng replied:
we can put a warning message

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCommon.java:
```diff
@@ -0,0 +1,247 @@
+        if (!this.isRoot) {
+            return;
+        }
+        CgroupUtils.writeFileByLine(CgroupUtils.getDir(this.dir, RELEASE_AGENT), command);
```
this operation should override the existing file, isn't it?

jerrypeng replied:
doesn't matter to cgroup config files, please see me previous comment

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCommon.java:
```diff
@@ -0,0 +1,247 @@
+        if (!this.isRoot) {
+            return null;
+        }
+        return CgroupUtils.readFileByLine(CgroupUtils.getDir(this.dir, RELEASE_AGENT)).get(0);
```
put a check for number of rows that are returned.

jerrypeng replied:
not necessary, see previous comment

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCommon.java:
```diff
@@ -0,0 +1,247 @@
+
+    @Override
+    public void setCgroupCloneChildren(boolean flag) throws IOException {
+        if (!getCores().keySet().contains(SubSystemType.cpuset)) {
```
It is not clear why this condition is present.

jerrypeng replied:
because its a config for only the subsystem cpuset

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCommon.java:
```diff
@@ -0,0 +1,247 @@
+        if (!getCores().keySet().contains(SubSystemType.cpuset)) {
+            return;
+        }
+        CgroupUtils.writeFileByLine(CgroupUtils.getDir(this.dir, CGROUP_CLONE_CHILDREN), flag ? "1" : "0");
```
writing a boolean into a file can go into utility function

jerrypeng replied:
not necessary, not generic enough

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCommon.java:
```diff
@@ -0,0 +1,247 @@
+
+        File file = new File(this.dir);
+        File[] files = file.listFiles();
+        if (files == null) {
```
this won't happen.

jerrypeng replied:
please look the function definition for listFiles().  The function can return null

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCommon.java:
```diff
@@ -0,0 +1,247 @@
+    public int hashCode() {
+        final int prime = 31;
+        int result = 1;
+        result = prime * result + (this.name != null ? this.name.hashCode() : 0);
```
you can use lombok plugin to generate these methods. The null checks you are doing in hashCode, should be done in equals method as well.

jerrypeng replied:
will correct

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupManager.java:
```diff
@@ -0,0 +1,207 @@
+        this.conf = conf;
+        this.rootDir = Config.getCgroupRootDir(this.conf);
+        if (this.rootDir == null) {
+            throw new RuntimeException("Check configuration file. The supervisor.cgroup.rootdir is missing.");
```
should be storm.supervisor.cgroup.rootdir is missing.

jerrypeng replied:
will fix

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupManager.java:
```diff
@@ -0,0 +1,207 @@
+
+        File file = new File(Config.getCgroupStormHierarchyDir(conf) + "/" + this.rootDir);
+        if (!file.exists()) {
+            LOG.error("{}/{} is not existing.", Config.getCgroupStormHierarchyDir(conf), this.rootDir);
```
you can simplify by using file.getPath()

jerrypeng replied:
will refactor

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupManager.java:
```diff
@@ -0,0 +1,207 @@
+            try {
+                memCore.setPhysicalUsageLimit(Long.valueOf(totalMem.longValue() * 1024 * 1024));
+            } catch (IOException e) {
+                throw new RuntimeException("Cannot set memory.limit_in_bytes! Exception: " + e);
```
use, e instead of + e

jerrypeng replied:
will change

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupUtils.java:
```diff
@@ -0,0 +1,128 @@
+
+    private static final Logger LOG = LoggerFactory.getLogger(CgroupUtils.class);
+
+    public static void deleteDir(String dir) {
```
There are already utility methods for doing this.

jerrypeng replied:
Yes we can use FileUtils.deleteDirectory() but we won't get the the helpful warning messages.

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupUtils.java:
```diff
@@ -0,0 +1,128 @@
+        return Utils.checkFileExists(CGROUP_STATUS_FILE);
+    }
+
+    public static List<String> readFileByLine(String fileDir) throws IOException {
```
fileDir can be renamed to filepath

abhishekagarwal87 replied:
readFileByLine can be replaced with Files.readlines(new File(fileDir)) available in guava and already added as dependency

jerrypeng replied:
will use

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/Device.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    @Override
+    public String toString() {
```
lot of boilerplate code can be done away with using lombok annotations.

jerrypeng replied:
we are not using lombok annotations in storm currently. Perhaps we can int the future, but perhaps @abhishekagarwal87 can file a jira to investigate whether we should use it

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/SubSystemType.java:
```diff
@@ -0,0 +1,52 @@
+    // net_cls,ns is not supported in ubuntu
+    blkio, cpu, cpuacct, cpuset, devices, freezer, memory, perf_event, net_cls, net_prio;
+
+    public static SubSystemType getSubSystem(String str) {
```
SubSystemType.valueOf(str) will do the exact same thing.

jerrypeng replied:
will change

zhuoliu commented on conf/defaults.yaml:
```diff
@@ -263,7 +263,7 @@ topology.state.checkpoint.interval.ms: 1000
 # Recommended range of 0-29 but no hard limit set.
 topology.priority: 29
-topology.component.resources.onheap.memory.mb: 128.0
+topology.component.resources.onheap.memory.mb: 256.0
```
Do we have a reason for doubling the default per-instance memory usage?

jerrypeng replied:
So system components (i.e. acker and writer) use more memory than 128.0mb. When I turned on cgroups, workers will just system components on them will get consistently killed by cgroups for using more memory than allowed

zhuoliu replied:
Should we have a separate default value for system components? Otherwise, users' topologies might use more memory than their previous impression. E.g., a WordCount example might need 6GB to start. Since this is an important setting, @revans2 , what is your opinion?

revans2 replied:
First of all RAS should be keeping scheduling ackers just like any other bolt.  The system will have some overhead, and so will the JVM itself. If we assume that the JVM will only ever use the amount of memory specified in the HEAP we are going to be in trouble when someone starts to use their entire heap.

I would say initially that we want a two configs that are for worker system overhead, both on heap and off heap.  Then we reserve that many MBs * number_of_slots from being used for scheduling.  Increasing the default does not solve this problem, except in the case where we are not using all of the heap.

zhuoliu commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupUtils.java:
```diff
@@ -0,0 +1,128 @@
+        LOG.debug("For CGroups - writing {} to {} ", strings, fileDir);
+        File file = new File(fileDir);
+        if (!file.exists()) {
+            LOG.error("{} is no existed", fileDir);
```
not => not

jerrypeng replied:
will change

zhuoliu commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupOperation.java:
```diff
@@ -0,0 +1,80 @@
+     * get the first hierarchy that has a certain list of subsystems isMounted
+     */
+    public Hierarchy getHierarchyWithSubSystems(List<SubSystemType> subSystems);
+
```
extra line

jerrypeng replied:
will delete

zhuoliu commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupManager.java:
```diff
@@ -0,0 +1,207 @@
+
+        // set upper limit to how much cpu can be used by all workers running on supervisor node.
+        // This is done so that some cpu cycles will remain free to run the daemons and other miscellaneous OS operations.
+        CpuCore supervisorRootCPU = (CpuCore)  this.rootCgroup.getCores().get(SubSystemType.cpu);
```
extra space

jerrypeng replied:
will remove

zhuoliu commented on storm-core/src/jvm/org/apache/storm/container/cgroup/core/DevicesCore.java:
```diff
@@ -0,0 +1,185 @@
+
+    public static final char ACCESS_READ_CH = 'r';
+    public static final char ACCESS_WRITE_CH = 'w';
+    public static final char ACCESS_CREATE_CH = 'm';
```
Can we make these final values private? Since they are only used in this class.

jerrypeng replied:
will refactor

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCenter.java:
```diff
@@ -0,0 +1,218 @@
+        Set<SubSystemType> subSystems = hierarchy.getSubSystems();
+        for (SubSystemType type : subSystems) {
+            if (this.getHierarchyWithSubSystem(type) != null) {
+                LOG.error("subSystem: {} is busy", type.name());
```
can you also log the hierarchy directory to which it is already attached.

jerrypeng replied:
will add

jerrypeng commented:
@abhishekagarwal87 Thanks for the detailed review! much appreciated!  I have addressed your comments
@zhuoliu thanks for your review as well.  I have address your comments as well

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/SystemOperation.java:
```diff
@@ -0,0 +1,75 @@
+            }
+            return output;
+        } catch (InterruptedException ie) {
+            throw new IOException(ie.toString());
```
ie can be wrapped instead of ie.toString()

jerrypeng replied:
will refactor

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/SystemOperation.java:
```diff
@@ -0,0 +1,75 @@
+
+    public static void umount(String name) throws IOException {
+        StringBuilder sb = new StringBuilder();
+        sb.append("umount ").append(name);
```
variable should be named dir.

jerrypeng replied:
will rename

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/SystemOperation.java:
```diff
@@ -0,0 +1,75 @@
+        sb.append("mount -t ")
+                .append(type)
+                .append(" -o ")
+                .append(data)
```
variable data can be renamed to options.

jerrypeng replied:
will rename

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/core/BlkioCore.java:
```diff
@@ -0,0 +1,226 @@
+        read, write, sync, async, total;
+
+        public static RecordType getType(String type) {
+            if (type.equals("Read")) {
```
(type == null) ? null : RecordType.valueOf(type.toLowerCase())

jerrypeng replied:
will change

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/core/CpuacctCore.java:
```diff
@@ -0,0 +1,71 @@
+        return result;
+    }
+
+    public enum StatType {
```
It is better to declare inner class,enums static.

jerrypeng replied:
Nested enum types are implicitly static.

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/core/CpusetCore.java:
```diff
@@ -0,0 +1,209 @@
+        int value = 0;
+        int start = 0;
+        boolean isHyphen = false;
+        for (char ch : output) {
```
can this be simplified to 
```
for (str in outputStr.split(",")) 
if (!str.contains("-"))
numList.add(Integer.valueOf(str))
else {
Integer start = Integer.valueOf(str.split("-")[0])
Integer end = Integer.valueOf(str.split("-")[1])
// Add from start to end to numList
}
```
It is based on my understanding of list format. Hope it covers the cases

d2r replied:
I agree this is not the most straightforward way of doing this in Java and
looks more C-like. 

I think I would be OK with leaving the code as-is, and maybe we could add a
comment explaining what this does for others since it is? We could clean it up
in follow-on work if it still is a problem.

```
/** 
 * Accepts a string of comma-separated values or ranges of values.  Specify a
 * value like '42' and a range like '12-17'.
 *
 * Example: '42,12-17,100' -> {42, 12, 13, 14, 15, 16, 17, 100}
 */
```

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/core/CpusetCore.java:
```diff
@@ -0,0 +1,209 @@
+            }
+        }
+
+        int[] nums = new int[numList.size()];
```
numList.toArray()

jerrypeng replied:
not that simple. numList is a List<Integer> while the function returns int[].

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/core/DevicesCore.java:
```diff
@@ -0,0 +1,185 @@
+
+        public Record(String output) {
+            if (output.contains("*")) {
+                System.out.println("Pre:" + output);
```
replace with logger

jerrypeng replied:
will replace

abhishekagarwal87 commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupUtils.java:
```diff
@@ -0,0 +1,119 @@
+            LOG.error("{} does not exist", filePath);
+            return;
+        }
+        try (FileWriter writer = new FileWriter(file, true);
```
append is set to true here. should it be set to false

jerrypeng replied:
doesn't make a different for the cgroup config files

abhishekagarwal87 commented:
Thanks @jerrypeng. This is quite a beast. Learnt a good deal about cgroups through this PR.

knusbaum commented on storm-core/src/clj/org/apache/storm/daemon/supervisor.clj:
```diff
@@ -344,6 +347,12 @@
    :download-lock (Object.)
    :stormid->profiler-actions (atom {})
+   :resource-isolation-manager (if (conf STORM-RESOURCE-ISOLATION-PLUGIN-ENABLE)
+                     (let [resource-isolation-manager (Utils/newInstance (conf STORM-RESOURCE-ISOLATION-PLUGIN))]
```
Indentation here is off.

jerrypeng replied:
will fix

knusbaum commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupManager.java:
```diff
@@ -0,0 +1,210 @@
+    }
+
+    /**
+     * User cfs_period & cfs_quota to control the upper limit use of cpu core e.g.
```
User -> Use

jerrypeng replied:
will fix

knusbaum commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupManager.java:
```diff
@@ -0,0 +1,210 @@
+            // No control of cpu usage
+            cpuCore.setCpuCfsQuotaUs(cpuCoreUpperLimit);
+        } else {
+            cpuCore.setCpuCfsPeriodUs(100000);
```
Is this supposed to be a number literal?

jerrypeng replied:
yes

knusbaum replied:
Good.

knusbaum commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupOperation.java:
```diff
@@ -0,0 +1,78 @@
+import java.util.Set;
+
+/**
+ * An interface to manage cgroups
```
This doc could use a little expansion. It's so vague as to be basically useless. The rest in this file look good.

jerrypeng replied:
will elaborate

knusbaum commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupUtils.java:
```diff
@@ -0,0 +1,119 @@
+
+    public static void deleteDir(String dir) {
+        File d = new File(dir);
+        if (d.exists()) {
```
Can we un-nest this so it's easier to follow? Something like: 
```
if (!d.exists()) {
    Log.warn("dir {} does not exist!", dir);
    return;
}
if (!d.isDirectory()) {
    throw new RuntimeException("dir " + dir + " is not a directory!");
}
if (!d.delete()) {
    throw new RuntimeException("Cannot delete dir " + dir);
}
```

jerrypeng replied:
will refactor

knusbaum commented:
+1

jerrypeng commented:
I upmerged

d2r commented on storm-core/src/jvm/org/apache/storm/Config.java:
```diff
@@ -2194,6 +2195,73 @@
+     * The amount of memory a worker can exceed its allocation before cgroup will kill it
+     */
+    @isPositiveNumber
+    public static String STORM_CGROUP_MEMORY_MB_LIMIT_TOLERANCE_MARGIN = "storm.cgroup.memory.limit.tolerance.margin.mb";
```
We should make the java variable match the config name string like the others.
`STORM_CGROUP_MEMORY_LIMIT_TOLERANCE_MARGIN_MB`.

d2r commented:
One comment about the new tolerance config name, other than that it looks OK.

jerrypeng commented:
@d2r thanks for your review I have addressed your comment

zhuoliu commented:
Looks good! +1

d2r commented on storm-core/src/jvm/org/apache/storm/Config.java:
```diff
@@ -2194,6 +2195,73 @@
+     * The amount of memory a worker can exceed its allocation before cgroup will kill it
+     */
+    @isPositiveNumber
+    public static String STORM_CGROUP_MEMORY_MB_LIMIT_TOLERANCE_MARGIN = "storm.cgroup.memory.mb.limit.tolerance.margin";
```
This is not exactly what I had in mind from my review comment. Now the Java name is consistent with the config name, but if we look at the other config definitions, there is also a consistency in the units of the value being appended to the name, like `topology.worker.max.heap.size.mb`. This PR does something different with `storm.worker.cgroup.memory.mb.limit` and `storm.cgroup.memory.mb.limit.tolerance.margin` where the unit is somewhere in the middle of the config.  I think it is better to be consistent with the other existing configs.

jerrypeng replied:
Ok will fix

d2r commented:
Travis CI [error](https://travis-ci.org/apache/storm/jobs/109239992#L1570) does not seem related. Created [STORM-1555](https://issues.apache.org/jira/browse/STORM-1555)

jerrypeng commented:
@dr2 thanks for review, I have addressed your comment

d2r commented:
+1 looks good. Thank you.

vesense commented:
LGTM
